neqo-http3/src/send_message.rs

```diff
@@ -0,0 +1,270 @@
+    }
+
+    pub fn send_body(&mut self, conn: &mut Connection, buf: &[u8]) -> Res<usize> {
+        qinfo!(
+            [self],
+            "send_request_body: state={:?} len={}",
+            self.state,
+            buf.len()
+        );
+        match self.state {
+            SendMessageState::Uninitialized
+            | SendMessageState::Initialized { .. }
+            | SendMessageState::SendingInitialMessage { .. } => Ok(0),
+            SendMessageState::SendingData => {
+                let available = usize::try_from(conn.stream_avail_send_space(self.stream_id)?)
+                    .unwrap_or(usize::max_value());
+                if available <= 2 {
+                    return Ok(0);
+                }
+                let to_send;
+                if available <= MAX_DATA_HEADER_SIZE_2_LIMIT {
+                    // 63 + 3
+                    to_send = min(min(buf.len(), available - 2), MAX_DATA_HEADER_SIZE_2);
+                } else if available <= MAX_DATA_HEADER_SIZE_3_LIMIT {
+                    // 16383 + 5
+                    to_send = min(min(buf.len(), available - 3), MAX_DATA_HEADER_SIZE_3);
+                } else if available <= MAX_DATA_HEADER_SIZE_5 {
+                    // 1073741823 + 9
+                    to_send = min(min(buf.len(), available - 5), MAX_DATA_HEADER_SIZE_5_LIMIT);
+                } else {
+                    to_send = min(buf.len(), available - 9);
+                }
+
+                qinfo!(
+                    [self],
+                    "send_request_body: available={} to_send={}.",
+                    available,
+                    to_send
+                );
+
+                let data_frame = HFrame::Data {
+                    len: to_send as u64,
+                };
+                let mut enc = Encoder::default();
+                data_frame.encode(&mut enc);
+                match conn.stream_send(self.stream_id, &enc) {
+                    Ok(sent) => {
+                        debug_assert_eq!(sent, enc.len());
+                    }
+                    Err(e) => return Err(Error::TransportError(e)),
+                }
+                match conn.stream_send(self.stream_id, &buf[..to_send]) {
+                    Ok(sent) => Ok(sent),
+                    Err(e) => Err(Error::TransportError(e)),
+                }
+            }
+            SendMessageState::Closed => Err(Error::AlreadyClosed),
+        }
+    }
+
+    pub fn is_sending_closed(&self) -> bool {
+        match self.state {
+            SendMessageState::Initialized { fin, .. }
+            | SendMessageState::SendingInitialMessage { fin, .. } => fin,
+            SendMessageState::SendingData => false,
+            _ => true,
+        }
+    }
+
+    pub fn done(&self) -> bool {
+        matches!(self.state, SendMessageState::Closed)
+    }
+
+    pub fn is_state_sending_data(&self) -> bool {
+        matches!(self.state, SendMessageState::SendingData)
+    }
+
+    fn ensure_response_encoded(
```

Comment:
Is this code generic?  That is, should it be response-specific or does it apply to requests also?  It looks like it.  Is this better named `encode_header`?

Reply:
ah, yes.it is generic.
it is not only headers. The server side may have data as well. That will be fix once the server side is fixed, but that is not the highest priority at the moment.